### ISSUE TITLE
s3.to_parquet(): weird behaviour when inserting existing partitions on a different location with mode='overwrite_partitions'

### ISSUE THREAD

**Issue by VashMKS** [awslabs/aws-data-wrangler#586]
While debugging one of our processes I came to notice the following behaviour:

## Setup
* I have a partitioned table with a single partition, let's say the values for the partition are x, y, z. The data in S3 looks like this:
```
s3://mybucket/
└── myprefix/
    ├── mypartition=x/<uuid1>.snappy.parquet
    ├── mypartition=y/<uuid2>.snappy.parquet
    └── mypartition=z/<uuid3>.snappy.parquet
```
* I *mistakenly* upsert (via `s3.to_parquet()` with `mode="overwrite_partitions"`) a new partition (`partition=t`) on a *different* prefix (`path="s3://otherprefix/"`). Now my table in Catalog still displays its location as `s3://mybucket/myprefix/` but the data looks like:
```
s3://mybucket/
├── myprefix/
│   ├── mypartition=x/<uuid1>.snappy.parquet
│   ├── mypartition=y/<uuid2>.snappy.parquet
│   └── mypartition=z/<uuid3>.snappy.parquet
└── otherprefix/
    └── mypartition=t/<uuid4>.snappy.parquet
```
* Upon finding the mistake, my immediate expectation was that performing the same operation (`s3.to_parquet()` with `mode="overwrite_partitions"`) for partition `t` with the correct prefix `myprefix` would fix it by:
  1. Removing the files from `otherprefix`
  2. Writting the data for partition `t` to `myprefix/mypartition=t/`
  3. Updating the Glue Catalog table so as to point the existing partition `t` to `myprefix/partition=t/`

It only did the second of those things, while issuing no warning or acknowledgement.

I only noticed because I double checked S3 and saw that while partition `t` had been correctly created in `myprefix/`, `otherprefix/` was still there. I though that maybe wrangler defaults to not deleting that partition since it's not physically overlapping with the new one and I might want to preserve the data for some reason (which IMO would be fine but still warrant some warning to prevent zombie files), so I went ahead and deleted it manually. I then noticed that Athena could not find the new data, which prompted me to manually check to which location the table partitions pointed to in S3 and of course they pointed to the old files in `otherprefix/` I'd just deleted.

To see if my expectations were unreasonable, I went back and read the [docs on `s3.to_parquet`](https://aws-data-wrangler.readthedocs.io/en/2.4.0-docs/stubs/awswrangler.s3.to_parquet.html) (BTW the [link to the tutorial in the docstring for the mode parameter](https://github.com/awslabs/aws-data-wrangler/blob/b58359199d49758a03a54854118ce5886fab382a/awswrangler/s3/_write_parquet.py#L298) is broken), related tutorials 4, 5 and 14, the [Glue Docs](https://docs.aws.amazon.com/glue/latest/dg/populate-data-catalog.html) and of course other issues but found no mention of this behaviour.

TL;DR: Upserting an existing partition value with `mode="overwrite_partitions"` for a partition where the physical files are located somewhere different to the path you indicate will cause the new files to be writen to S3 but not added to the table and the old files to not be deleted. I find this confusing at best and a hazard for data integrity at worst.

## Discussion & Workarounds
After putting some thought into it, I'm confident this is either a bug, intended but in need of documentation surrounding the interaction so that we can build around it, or maybe underspecification on the Data Catalog side of things, but I'm not sure which one it is.

Maybe Data Catalog simply does not contemplate the use case where one table's data has multiple prefixes in S3, though in that case it should issue an exception when trying to partially update a table with data from a different location. Since it does not, I'll assume it's fine to have data from several locations be part of the same table (though probably not great fun).

If I were to treat this as a bug and point to somewhere in the code I would note that `to_parquet` calls `_to_dataset` which calls `_to_partitions` when called with `dataset=True` and some `partition_cols`. This function then does the following:

```python
def _to_partitions(...):
  for keys, subgroup in df.groupby(by=partition_cols, observed=True):
          subgroup = subgroup.drop(partition_cols, axis="columns")
          keys = (keys,) if not isinstance(keys, tuple) else keys
          subdir = "/".join([f"{name}={val}" for name, val in zip(partition_cols, keys)])
          prefix: str = f"{path_root}{subdir}/"
          if mode == "overwrite_partitions":
              delete_objects(
                  path=prefix,
                  use_threads=use_threads,
                  boto3_session=boto3_session,
                  s3_additional_kwargs=func_kwargs.get("s3_additional_kwargs"),
              )
  ...
```

When given `mode="overwrite_partitions"`, `_to_partitions` is naive and checks only the given path (which is the one we are trying to write the new data to) and removes an existing partition if there. However it does not check if the Catalog already has an entry for said partition value, which as we saw could hypothetically be located somewhere else. This is the change I'd suggest at first glance, though it does not seem great performance-wise.

Finally, here's a workaround to all of this that just came to mind, though for often modified, highly partitioned tables in that particular use case might be expensive:
1. Delete the old files manually after insertion of the new ones
2. Run a crawler to update the partitions

I haven't put much work into investigating the issue further since I'm relatively new to both awswrangler and the Data Catalog and maybe there's something really obvious that our small in-house team missed. I think that this could cause some serious silent bugs so I decided to raise awareness. Hopefully I'm just missing something. If not, at least we might get some interesting discussion going.

PS: Cheers for the library, it absolutely rocks. We started building our own version of pandas meets AWS about a year ago but recently found this and will definitely be porting our codebases, this will save us TONS of work. Once I'm more familiar with the whole thing and get some time in my hands I'd love to contribute.

**Comment by jaidisido** [comment]
Thanks for this detailed explanation @VashMKS. One change we can make is to simply raise an exception when the existing Glue table prefix does not match the one specified in the `s3.to_parquet` call. It would have prevented you from making the mistake in the first place. We can address this in the next release (2.6.0).

**Comment by jaidisido** [comment]
Covered in release [2.6.0](https://github.com/awslabs/aws-data-wrangler/releases/tag/2.6.0)
